### PR TITLE
build: check FP16 build condition correctly

### PR DESCRIPTION
**Pull request by tomoaki0705**
### This pullrequest changes

  * use GNUC_MINOR in correct place to check the version of GCC
  * guard using ifdef correctly in case of various configuration
  * use v_float16x4 correctly by including the right header file
  * check FP16 compiler support using cmake correctly

### CODE REVIEW

File: cmake/OpenCVCompilerOptions.cmake [hunk 1]
```diff
@@ -377,10 +377,15 @@ if(MSVC)
 endif()
 
 if(NOT OPENCV_FP16_DISABLE)
+  if(ARM)
+    set(FP16_OPTION "-mfpu=neon-fp16")
+  elseif(X86 OR X86_64)
+    set(FP16_OPTION "-mf16c")
+  endif()
```
Comment (alalek) [inline cmake/OpenCVCompilerOptions.cmake]: In case of successful check you should update compiler flags with these options.

Comment (tomoaki0705) [inline cmake/OpenCVCompilerOptions.cmake]: Thank you @alalek for your time and reviewing my PR.
I updated the PR and I had to change this ```try_compile``` to ```try_run```
I'd like to hear your opinion about that point.

Best.

Comment (alalek) [inline cmake/OpenCVCompilerOptions.cmake]: I believe `try_compile` is preferable here. `try_run` doesn't work in case of cross-compilation.

Comment (tomoaki0705) [inline cmake/OpenCVCompilerOptions.cmake]: @alalek , thank you for your review and suggestion.
I switched back to ```try_compile```
I was facing a problem that when the processor doesn't support FP16, the program was crashing with ``illegal instruction``
I checked and this happened even I disabled the ```CV_FP16```
I then realized that the generated binary had some AVX instructions.
```
~/opencv-fork/build$ cd modules/core
~/opencv-fork/build/modules/core$ make test/test_intrin.cpp.s
~/opencv-fork/build/modules/core$ grep vadd CMakeFiles/opencv_test_core.dir/test/test_intrin.cpp.s
        vaddps  %xmm0, %xmm1, %xmm1
        vaddps  %xmm2, %xmm3, %xmm0
        vaddps  %xmm1, %xmm0, %xmm0
```
Those ```vaddps``` instructions were not from the ```FP16``` related part.
So I let the FP16 feature rely on AVX flag, too.

**Comment by alalek** [comment]
:+1:

**Pull request merged**